### PR TITLE
feat(crypto-nodejs) Implement missing APIs

### PR DESCRIPTION
**Pull request by Hywan**
Implementation of https://github.com/matrix-org/matrix-rust-sdk/issues/699.

The work has started in #675. This PR is the sequel of it.

## Progress

The methods that are needed for basic encryption/decryption are:

- [x] `OlmMachine::new()` -> `OlmMachine::initialize()` (since JavaScript cannot have async constructor, we use the factory pattern)
- [x] `OlmMachine::receive_sync_changes()`
- [x] `OlmMachine::outgoing_requests()`
- [x] `OlmMachine::mark_request_as_sent()`
- [x] `OlmMachine::decrypt_room_event()`
- [x] `OlmMachine::update_tracked_users()`
- [x] `OlmMachine::get_missing_sessions()`
- [x] `OlmMachine::share_room_key()`
- [x] `OlmMachine::encrypt_raw()`

Bonus:

- [x] `OlmMachine::user_id()`
- [x] `OlmMachine::device_id()`
- [x] `OlmMachine::identity_keys()`

~Note: This PR includes https://github.com/matrix-org/matrix-rust-sdk/pull/723 temporarily.~

Other tasks than the binding itself:

- [x] Test suites are written to cover the entire API
- [x] CI has been updated to run the test
- [x] Documentation can be generated
- Documentation is online (will be addressed in another PR to keep the scope of this one small)
- [x] `README.md` is written (and helpful)

### CODE REVIEW

File: crates/matrix-sdk-crypto-nodejs/src/machine.rs [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    // We could be tempted to use `requests::OutgoingRequests` as its
+    // a type alias for this giant `Either` chain. But `napi` won't
+    // unfold it properly into a valid TypeScript definition, so…
+    // let's copy-paste :-(.
+    #[napi]
+    pub async fn outgoing_requests(
+        &self,
+    ) -> Result<
+        Vec<
+            Either<
```
Comment (poljar) [inline crates/matrix-sdk-crypto-nodejs/src/machine.rs:107]: Can we at least type alias this monstrosity?

Comment (jplatte) [reply crates/matrix-sdk-crypto-nodejs/src/machine.rs:107]: See comment above the function..

Comment (poljar) [reply crates/matrix-sdk-crypto-nodejs/src/machine.rs:107]: Ah missed that it was talking about a type alias, considering that `OutgoingRequests` is not a type alias.

File: crates/matrix-sdk-crypto-nodejs/src/machine.rs [hunk 2]
```diff
@@ -109,20 +112,15 @@ impl OlmMachine {
         &self,
     ) -> Result<
         Vec<
-            Either<
+            Either5<
```
Comment (jplatte) [inline crates/matrix-sdk-crypto-nodejs/src/machine.rs:115]: Oh, five is the biggest so you still need (direct) nesting? This seems like sth. that should be fixed in napi 🤔

Comment (Hywan) [reply crates/matrix-sdk-crypto-nodejs/src/machine.rs:115]: It is yes. I will work on a patch to support up to 26 with the help of macros. We need to update 2 files: [`either.rs`](https://github.com/napi-rs/napi-rs/blob/d6c755c4eb8b2a56e226866d9a595ad2d712ebcc/crates/napi/src/bindgen_runtime/js_values/either.rs) and [`typegen.rs`](https://github.com/napi-rs/napi-rs/blob/d6c755c4eb8b2a56e226866d9a595ad2d712ebcc/crates/backend/src/typegen.rs).

Comment (Hywan) [reply crates/matrix-sdk-crypto-nodejs/src/machine.rs:115]: Done, https://github.com/napi-rs/napi-rs/pull/1200.

File: crates/matrix-sdk-crypto-nodejs/src/machine.rs [hunk 3]
```diff
@@ -62,8 +62,10 @@ impl OlmMachine {
     /// asynchronous. Please use the `finalize` method.
     #[napi(constructor)]
     #[allow(clippy::new_ret_no_self)]
-    pub fn new() -> napi::Error {
-        napi::Error::from_reason("To build an `OldMachine`, please use the `initialize` method")
+    pub fn new() -> Result<(), napi::Error> {
```
Comment (jplatte) [inline crates/matrix-sdk-crypto-nodejs/src/machine.rs:65]:
```suggestion
    pub fn new() -> napi::Result<()> {
```

Comment (Hywan) [reply crates/matrix-sdk-crypto-nodejs/src/machine.rs:65]: Oh sorry, I misread your diff. Yeah, I should probably switch to `napi::Result` :-).

File: crates/matrix-sdk-crypto-nodejs/.gitignore [hunk 4]
```diff
@@ -2,4 +2,5 @@
 /package-lock.json
 /index.js
 /index.d.ts
-/matrix-sdk-crypto.*.node
\ No newline at end of file
+/matrix-sdk-crypto.*.node
+/docs/*
```
Comment (jplatte) [inline crates/matrix-sdk-crypto-nodejs/.gitignore:6]: Missing final newline ;)

Comment (Hywan) [reply crates/matrix-sdk-crypto-nodejs/.gitignore]: ;-)

File: crates/matrix-sdk-crypto-nodejs/Cargo.toml [hunk 5]
```diff
@@ -22,13 +22,18 @@ crate-type = ["cdylib"]
...
-napi-derive = "2.4.1"
+vodozemac = {  git = "https://github.com/matrix-org/vodozemac/", rev = "d0e744287a14319c2a9148fef3747548c740fc36" }
+napi = { git = "https://github.com/Hywan/napi-rs", branch = "feat-either-n-up-to-26", default-features = false, features = ["napi6", "tokio_rt"] }
+napi-derive = { git = "https://github.com/Hywan/napi-rs", branch = "feat-either-n-up-to-26" }
+serde_json = "1.0.79"
+http = "0.2.6"
+tracing-subscriber = { version = "0.3", default-features = false, features = ["tracing-log", "time", "smallvec", "fmt"], optional = true }
 
 [build-dependencies]
 napi-build = "2.0.0"
```
Comment (jplatte) [inline crates/matrix-sdk-crypto-nodejs/Cargo.toml:39]: Missing final newline.

File: crates/matrix-sdk-crypto-nodejs/README.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+If you want to enable [tracing](https://tracing.rs), i.e. to get the
+logs, you should re-compile the extension with the `tracing` feature
+turned on:
+
+```sh
+$ npm run build -- --features tracing
+```
+
+Now, you can use the `RUST_LOG` environment variable to tweak the log filtering, such as:
```
Comment (jplatte) [inline crates/matrix-sdk-crypto-nodejs/README.md:103]: This only works with `tracing-subscriber`s `env-filter` feature turned on, but you didn't enable it.

File: crates/matrix-sdk-crypto-nodejs/src/identifiers.rs [hunk 7]
```diff
@@ -43,13 +46,13 @@ impl UserId {
     /// A historical user ID is one that doesn't conform to the latest
     /// specification of the user ID grammar but is still accepted
     /// because it was previously allowed.
-    #[napi(getter, js_name = "isHistorical")]
```
Comment (jplatte) [inline crates/matrix-sdk-crypto-nodejs/src/identifiers.rs:46]: Why are these `js_name`s removed (and the getter attribute here, too)?

**Review by jplatte — COMMENTED** [review]
Incomplete review, will continue later.

File: crates/matrix-sdk-crypto-nodejs/src/identifiers.rs (hunk 7 above)
Comment (Hywan) [reply crates/matrix-sdk-crypto-nodejs/src/identifiers.rs:46]: `napi-rs` automatically transforms `foo_bar` to `fooBar`, so using all those `js_name`s is now useless.

Also, about the `getter` being removed, I don't feel like `is*` functions are getters, but more real functions semantically speaking.

File: crates/matrix-sdk-crypto-nodejs/src/machine.rs [hunk 8]
```diff
@@ -0,0 +1,439 @@
...
+    ///
+    /// # Arguments
+    ///
+    /// * `users`, an array over user IDs that should be marked for tracking.
+    #[napi]
+    pub async fn update_tracked_users(&self, users: Vec<&identifiers::UserId>) {
+        let users =
+            users.into_iter().map(|user| user.inner.clone()).collect::<Vec<ruma::OwnedUserId>>();
+
+        self.inner.update_tracked_users(users.iter().map(AsRef::as_ref)).await;
```
[the remark this answers is not in the document]
Comment (jplatte) [reply crates/matrix-sdk-crypto-nodejs/src/machine.rs:270]: I would be surprised if there was no way around this, but whatever. Can improve it in a follow-up PR.

**Review by jplatte — APPROVED** [review]

File: crates/matrix-sdk-crypto-nodejs/src/responses.rs [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+impl TryFrom<(RequestType, http::Response<Vec<u8>>)> for OwnedResponse {
+    type Error = napi::Error;
+
+    fn try_from(
+        (request_type, response): (RequestType, http::Response<Vec<u8>>),
+    ) -> Result<Self, Self::Error> {
+        match request_type {
+            RequestType::KeysUpload => {
+                KeysUploadResponse::try_from_http_response(response).map(Into::into)
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk-crypto-nodejs/src/responses.rs:84]: It will probably increase compilation time for a small code reduction.

**Pull request merged**